File: packages/g-copilot-proxy/app/auth/middleware.py
"""Authentication middleware."""

import logging
import time
from typing import Optional
from fastapi import Request, HTTPException, status
from fastapi.security import HTTPBearer, HTTPAuthorizationCredentials

from app.auth.config import get_auth_config

logger = logging.getLogger(__name__)
security = HTTPBearer(auto_error=False)


async def get_api_key(
    request: Request,
    credentials: Optional[HTTPAuthorizationCredentials] = None,
) -> str:
    """
    Get API key from stored OAuth credentials.

    GitHub Copilot requires OAuth device flow authentication.
    Personal access tokens are not supported.

    Raises HTTPException if credentials are not found or expired.
    """
    config = get_auth_config()
    stored_creds = config.get_stored_credentials()

    if not stored_creds:
        raise HTTPException(
            status_code=status.HTTP_401_UNAUTHORIZED,
            detail="No OAuth credentials found. Please authenticate at /auth/login",
        )

    access_token = stored_creds.get("access")
    if not access_token:
        raise HTTPException(
            status_code=status.HTTP_401_UNAUTHORIZED,
            detail="Invalid stored credentials. Please re-authenticate at /auth/login",
        )

    # Check if token is expired
    expires = stored_creds.get("expires")
    if expires:
        if time.time() * 1000 > expires:
            logger.warning("OAuth token expired. Please re-authenticate at /auth/login")
            raise HTTPException(
                status_code=status.HTTP_401_UNAUTHORIZED,
                detail="Token expired. Please re-authenticate at /auth/login",
            )

    return access_token


async def optional_auth(
    request: Request,
    credentials: Optional[HTTPAuthorizationCredentials] = None,
) -> Optional[str]:
    """
    Optional authentication - returns API key if available, None otherwise.
    Used for endpoints that work with or without auth.
    """
    try:
        return await get_api_key(request, credentials)
    except HTTPException:
        return None

File: packages/g-copilot-proxy/src/auth/middleware.ts
/**
 * Authentication middleware
 * Validates Bearer tokens or uses managed mode credentials
 */

import type { Context, Next } from "hono";
import type { Config } from "../config.js";
import { getLogger } from "../logger.js";
import { CredentialStorage } from "./storage.js";

const logger = getLogger();

/**
 * Authentication context attached to requests
 */
export interface AuthContext {
	accessToken: string;
	enterpriseUrl?: string;
	source: "bearer" | "managed";
}

/**
 * Create authentication middleware
 */
export function createAuthMiddleware(config: Config) {
	const storage = new CredentialStorage(config.DATA_DIR);

	return async (c: Context, next: Next) => {
		// In passthrough mode, look for Bearer token
		if (config.AUTH_MODE === "passthrough") {
			const authHeader = c.req.header("Authorization");
			if (authHeader?.startsWith("Bearer ")) {
				const token = authHeader.slice(7);
				c.set("auth", {
					accessToken: token,
					source: "bearer",
				});
				return next();
			}
			return c.json({ error: "Missing or invalid Authorization header" }, 401);
		}

		// In managed mode, use stored credentials
		const credentials = await storage.load();
		if (!credentials) {
			return c.json(
				{
					error: "No OAuth credentials found. Please authenticate at /auth/login",
				},
				401,
			);
		}

		// Check if credentials are expired
		const now = Date.now();
		const buffer = 5 * 60 * 1000; // 5 minute buffer

		if (now > credentials.expiresAt - buffer) {
			logger.warn("OAuth token expired. Please re-authenticate at /auth/login");
			return c.json(
				{
					error: "Token expired. Please re-authenticate at /auth/login",
				},
				401,
			);
		}

		c.set("auth", {
			accessToken: credentials.accessToken,
			enterpriseUrl: credentials.enterpriseUrl,
			source: "managed",
		});

		await next();
	};
}

/**
 * Helper to get auth context from request
 */
export function getAuthContext(c: Context): AuthContext {
	const auth = c.get("auth");
	if (!auth) {
		throw new Error("Auth context not found");
	}
	return auth;
}
